domainservice/src/main/kotlin/com/makarytskyi/rentcar/bpp/NatsInitializerBeanPostProcessor.kt

```diff
@@ -0,0 +1,61 @@
+package com.makarytskyi.rentcar.bpp
+
+import com.google.protobuf.GeneratedMessage
+import com.makarytskyi.rentcar.controller.nats.NatsController
+import io.nats.client.Connection
+import io.nats.client.Dispatcher
+import io.nats.client.MessageHandler
+import org.springframework.beans.factory.config.BeanPostProcessor
+import org.springframework.core.Ordered
+import org.springframework.core.annotation.Order
+import org.springframework.stereotype.Component
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.toMono
+
+@Component
+@Order(Ordered.LOWEST_PRECEDENCE)
+class NatsInitializerBeanPostProcessor(
+    private val dispatcher: Dispatcher,
+    private val connection: Connection
+) : BeanPostProcessor {
+
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any {
+        if (bean is NatsController<*, *>) processBean(bean)
+        return bean
+    }
+
+    private fun <RequestT : GeneratedMessage, ResponseT : GeneratedMessage> processBean(
+        bean: NatsController<RequestT, ResponseT>
+    ) {
+        val handler = MessageHandler { message ->
+            bean.parser.toMono()
+                .map { it.parseFrom(message.data) }
+                .flatMap { bean.handle(it) }
+                .onErrorResume { onParsingError(it, bean.defaultResponse) }
+                .subscribe {
+                    connection.publish(message.replyTo, it.toByteArray())
+                }
+        }
+        dispatcher.subscribe(bean.subject, bean.queueGroup, handler)
+    }
+
+    private fun <ResponseT : GeneratedMessage> onParsingError(
```

Comment:
We could get another exception, not only parsing error.
Suppose, your nats handler invokes a service, that invokes some infra staff, for example mongo.
Mongo could be down, so by the reverse chain you get some connectivity error.

I suggest to rename the function to `onUnexpectedError`